File: app/src/main/java/com/carpinelli/testeandroid/util/Mask.java
package com.carpinelli.testeandroid.util;

import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;

public class Mask {


    public static TextWatcher insertPhoneMask(final EditText ediTxt) {
        return new TextWatcher() {
            boolean isUpdating;
            String old = "";
            String mask9 = "(##)#####-####";
            String mask8 = "(##)####-####";
            String mask;

            public void onTextChanged(CharSequence s, int start, int before, int count) {

                String phone = ediTxt.getText().toString().trim().replaceAll("[^\\d.]", "");

                if (phone.length() < 11) {
                    mask = mask8;
                } else {
                    mask = mask9;
                }

                String str = s.toString().replaceAll("[.]", "").replaceAll("[-]", "")
                        .replaceAll("[/]", "").replaceAll("[(]", "")
                        .replaceAll("[)]", "");

                if (ediTxt.length() > 9) {
                    if (str.substring(0, 1).equals("0")) {
                        str = str.substring(1, str.length());
                    }
                }
                String mascara = "";
                if (isUpdating) {
                    old = str;
                    isUpdating = false;
                    return;
                }
                int i = 0;
                for (char m : mask.toCharArray()) {
                    if (m != '#' && str.length() != old.length()) {
                        mascara += m;
                        continue;
                    }
                    try {
                        mascara += str.charAt(i);
                    } catch (Exception e) {
                        break;
                    }
                    i++;
                }
                isUpdating = true;
                ediTxt.setText(mascara);
                ediTxt.setSelection(mascara.length());
            }

            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            public void afterTextChanged(Editable s) {
            }
        };
    }

}

File: app/src/main/java/com/vctapps/santanderchallenge/core/presentation/Mask.kt
package com.vctapps.santanderchallenge.core.presentation

import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText

object Mask {

    fun getTelNumberMask(edittext: EditText): TextWatcher{
        return object : TextWatcher{
            var oldText = ""
            var isUpdating = false

            override fun afterTextChanged(p0: Editable?) {}

            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}

            override fun onTextChanged(text: CharSequence?, p1: Int, p2: Int, p3: Int) {
                val textClear = removeMask(text.toString())

                val mask = if(textClear.length == 10) "(##) ####-####" else "(##) #####-####"

                var textWithMask = ""

                if (isUpdating) {
                    oldText = textClear
                    isUpdating = false
                    return
                }
                var i = 0

                for (m in mask.toCharArray()) {
                    if (m != '#' && textClear.length > oldText.length) {
                        textWithMask += m
                        continue
                    }
                    try {
                        textWithMask += textClear.get(i)
                    } catch (e: Exception) {
                        break
                    }

                    i++
                }

                isUpdating = true
                edittext.setText(textWithMask)
                edittext.setSelection(textWithMask.length)
            }
        }
    }

    fun removeMask(s: String): String {
        return s.replace("[.]".toRegex(), "")
                .replace("[-]".toRegex(), "")
                .replace("[/]".toRegex(), "")
                .replace("[(]".toRegex(), "")
                .replace("[ ]".toRegex(), "")
                .replace("[:]".toRegex(), "")
                .replace("[)]".toRegex(), "")
    }

}